# Challenge 1 
## Prueba de instalación NLTK

In [1]:
from nltk.corpus import brown

In [2]:
brown.words()[0:10]

['The',
 'Fulton',
 'County',
 'Grand',
 'Jury',
 'said',
 'Friday',
 'an',
 'investigation',
 'of']

In [3]:
from nltk import sent_tokenize, word_tokenize

In [4]:
>>> text = 'Ironhack is a Global Tech School ranked num 2 worldwide. 

Our mission is to help people transform their careers and join a thriving community of tech professionals that love what they do. This ideology is reflected in our teaching practices, which consist of a nine-weeks immersive programming, UX/UI design or Data Analytics course as well as a one-week hiring fair aimed at helping our students change their career and get a job straight after the course. We are present in 8 countries and have campuses in 9 locations - Madrid, Barcelona, Miami, Paris, Mexico City,  Berlin, Amsterdam, Sao Paulo and Lisbon.'

/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/IPython/core/inputtransformer2.py:481: UserWarning: `make_tokens_by_line` received a list of lines which do not have lineending markers ('\n', '\r', '\r\n', '\x0b', '\x0c'), behavior will be unspecified
  warnings.warn("`make_tokens_by_line` received a list of lines which do not have lineending markers ('\\n', '\\r', '\\r\\n', '\\x0b', '\\x0c'), behavior will be unspecified")


In [5]:
sent_tokenize(text)

['Ironhack is a Global Tech School ranked num 2 worldwide.',
 'Our mission is to help people transform their careers and join a thriving community of tech professionals that love what they do.',
 'This ideology is reflected in our teaching practices, which consist of a nine-weeks immersive programming, UX/UI design or Data Analytics course as well as a one-week hiring fair aimed at helping our students change their career and get a job straight after the course.',
 'We are present in 8 countries and have campuses in 9 locations - Madrid, Barcelona, Miami, Paris, Mexico City,  Berlin, Amsterdam, Sao Paulo and Lisbon.']

In [6]:
word_tokenize(text)

['Ironhack',
 'is',
 'a',
 'Global',
 'Tech',
 'School',
 'ranked',
 'num',
 '2',
 'worldwide',
 '.',
 'Our',
 'mission',
 'is',
 'to',
 'help',
 'people',
 'transform',
 'their',
 'careers',
 'and',
 'join',
 'a',
 'thriving',
 'community',
 'of',
 'tech',
 'professionals',
 'that',
 'love',
 'what',
 'they',
 'do',
 '.',
 'This',
 'ideology',
 'is',
 'reflected',
 'in',
 'our',
 'teaching',
 'practices',
 ',',
 'which',
 'consist',
 'of',
 'a',
 'nine-weeks',
 'immersive',
 'programming',
 ',',
 'UX/UI',
 'design',
 'or',
 'Data',
 'Analytics',
 'course',
 'as',
 'well',
 'as',
 'a',
 'one-week',
 'hiring',
 'fair',
 'aimed',
 'at',
 'helping',
 'our',
 'students',
 'change',
 'their',
 'career',
 'and',
 'get',
 'a',
 'job',
 'straight',
 'after',
 'the',
 'course',
 '.',
 'We',
 'are',
 'present',
 'in',
 '8',
 'countries',
 'and',
 'have',
 'campuses',
 'in',
 '9',
 'locations',
 '-',
 'Madrid',
 ',',
 'Barcelona',
 ',',
 'Miami',
 ',',
 'Paris',
 ',',
 'Mexico',
 'City',
 ',',
 '

# Challenge 2


In [7]:
import numpy as np
import re

In [8]:
path_url_1 = 'https?\:\/{2}[a-z]*\.?[a-zA-Z0-9@:%._\+~#=]*\.[a-zA-Z0-9()]*\b?[-a-zA-Z0-9()@:%_\+.~#?&//=]*' # con http 
#path_url_2 = '[-a-zA-Z0-9@:%._\+~#=]*\.[a-zA-Z0-9()]*\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)' # sin http
path_num = '[0-9]'
path_car = '\W'

In [9]:
# función de limpieza de strings

def clean_up(s):
    """
    Cleans up numbers, URLs, and special characters from a string.

    Args:
        s: The string to be cleaned up.

    Returns:
        A string that has been cleaned up.
    """
    s = re.sub(path_url_1, " ", s).lower()
    s = re.sub(path_num, " ",s)
    s = re.sub(path_car," ",s)
    return s

# Prueba
#prueba = '''@Ironhack's-#Q website 776-is http://ironhack.com [(2018)]")'''
#clean_up(prueba)    

In [10]:
# Tokenization

def tokenize(s):
    return word_tokenize(s)

In [11]:
# https://www.datacamp.com/community/tutorials/stemming-lemmatization-python

# Stemming & Lemmatizing
from nltk.stem import PorterStemmer
from nltk.stem import SnowballStemmer
from nltk.stem import WordNetLemmatizer

def stem_and_lemmatize(word_list):
    porter, lemma = PorterStemmer(),WordNetLemmatizer()  # inicializamos el stemmer y el lemmatizer
    lemm_stemm = [porter.stem(lemma.lemmatize(i)) for i in word_list] # Stemming -> Lemmatization
    return lemm_stemm

# PRUEBA
#word_list = ['hours',"friend", "friendship", "friends", "friendships","stabil","destabilize","misunderstanding","railroad","moonlight","football"]
#stem_and_lemmatize(word_list)


In [12]:
# Stop words
# https://www.datacamp.com/community/tutorials/text-analytics-beginners-nltk

from nltk.corpus import stopwords

stop_words = set(stopwords.words('english'))

def remove_stopwords(lst_stem): # Si la letra no esta en la lista de stopwords, no se agrega a la nueva lista.
    filtered_list= []
    for i in lst_stem:
        if i not in stop_words:
            filtered_list.append(i)
            
    return filtered_list

# Prueba
#word_list = ['a','hours',"friend", "friendship", "friends", "friendships","stabil","destabilize","misunderstanding","railroad","moonlight","football"]
#remove_stopwords(word_list)

# Challenge 3

In [13]:
# Leemos el csv
import zipfile
import pandas as pd

x = zipfile.ZipFile('Sentiment140.csv.zip')
df = pd.read_csv(x.open('Sentiment140.csv'))

In [14]:
# Agarramos un muestra del df de 20k
sample = df.sample(n=20000, random_state = 42).reset_index(drop = True) 
sample.head(3)

target          id                          date      flag  \
0       0  2200003196  Tue Jun 16 18:18:12 PDT 2009  NO_QUERY   
1       0  1467998485  Mon Apr 06 23:11:14 PDT 2009  NO_QUERY   
2       0  2300048954  Tue Jun 23 13:40:11 PDT 2009  NO_QUERY   

              user                                               text  
0  LaLaLindsey0609             @chrishasboobs AHHH I HOPE YOUR OK!!!   
1      sexygrneyes  @misstoriblack cool , i have no tweet apps  fo...  
2       sammydearr  @TiannaChaos i know  just family drama. its la...

In [15]:
# Aplicamos las funciones de limpieza del challenge 2 a cada row de la columna 'text'

sample['text_processed'] = sample['text'].apply(clean_up).apply(tokenize).apply(stem_and_lemmatize).apply(remove_stopwords)


In [16]:
sample.head(5)

target          id                          date      flag  \
0       0  2200003196  Tue Jun 16 18:18:12 PDT 2009  NO_QUERY   
1       0  1467998485  Mon Apr 06 23:11:14 PDT 2009  NO_QUERY   
2       0  2300048954  Tue Jun 23 13:40:11 PDT 2009  NO_QUERY   
3       0  1993474027  Mon Jun 01 10:26:07 PDT 2009  NO_QUERY   
4       0  2256550904  Sat Jun 20 12:56:51 PDT 2009  NO_QUERY   

              user                                               text  \
0  LaLaLindsey0609             @chrishasboobs AHHH I HOPE YOUR OK!!!    
1      sexygrneyes  @misstoriblack cool , i have no tweet apps  fo...   
2       sammydearr  @TiannaChaos i know  just family drama. its la...   
3      Lamb_Leanne  School email won't open  and I have geography ...   
4      yogicerdito                             upper airways problem    

                                      text_processed  
0                     [chrishasboob, ahhh, hope, ok]  
1            [misstoriblack, cool, tweet, app, razr]  
2  [tiannachao, know, famili, drama, lame, hey, n...  
3  [school, email, open, geographi, stuff, revis,...  
4                           [upper, airway, problem]

In [17]:
# Queremos conocer las palabras más frecuentes en el sample de tweets que tenemos.

from nltk import FreqDist 
lst_words = []
for i in sample['text_processed']:
    lst_words = lst_words+i

In [18]:
freq = FreqDist(lst_words)
most_common = freq.most_common(5000)
keys_common = [i[0] for i in most_common]  
keys_common[:10]

['go', 'wa', 'get', 'day', 'good', 'thi', 'work', 'like', 'love', 'quot']

In [37]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer
def find_features(tweets):
    features = {}
    for i in keys_common:
        features[i] = (i in tweets)
        
    s = SentimentIntensityAnalyzer().polarity_scores(" ".join(tweets))
    if s['pos'] > 0.3:
        s = True
    else:
        s = False
        
    return (features, s)
    return features

In [38]:
features = [find_features(i) for i in sample.text_processed]

In [29]:
from nltk import NaiveBayesClassifier

In [39]:
#Dividimos la base en  train y test
train = features[:10000]
test = features[10000:]

In [44]:
# Entrenamos los datos de train
classifier = NaiveBayesClassifier.train(train)

In [48]:
print(classifier.show_most_informative_features(15))

Most Informative Features
                 perfect = True             True : False  =     25.7 : 1.0
                   laugh = True             True : False  =     22.4 : 1.0
                  hahaha = True             True : False  =     19.8 : 1.0
                      xd = True             True : False  =     18.9 : 1.0
                   relax = True             True : False  =     17.2 : 1.0
                  talent = True             True : False  =     16.9 : 1.0
                    luck = True             True : False  =     16.1 : 1.0
                 comfort = True             True : False  =     12.5 : 1.0
                    hero = True             True : False  =     12.5 : 1.0
                   grand = True             True : False  =     12.5 : 1.0
                    love = True             True : False  =     12.1 : 1.0
                     win = True             True : False  =     12.0 : 1.0
                   great = True             True : False  =     11.2 : 1.0

In [49]:
nltk.classify.accuracy(classifier, test)

0.8432

Como se obtuvo una exactitud del 84% entonces se puede decir que el modelo está haciendo un buen trabajo.